### ISSUE TITLE
No more noSuchMethod (in dart2js/flutter production code)

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#28525]
**Warning**: Controversial.

## Overview

I'd like a flag, when both compiling to AOT for Flutter and dart2js for web to _not_ support `noSuchMethod` at all. For example, here is some code emitted via dart2js today within Angular:

```js
call$1: function($0) {
  return this.noSuchMethod$1(this, H.createInvocationMirror("call", "call$1", 0, [$0], []));
},
call$2: function($0, $1) {
  return this.noSuchMethod$1(this, H.createInvocationMirror("call", "call$2", 0, [$0, $1], []));
},
call$0: function() {
  return this.noSuchMethod$1(this, H.createInvocationMirror("call", "call$0", 0, [], []));
},
call$3: function($0, $1, $2) {
  return this.noSuchMethod$1(this, H.createInvocationMirror("call", "call$3", 0, [$0, $1, $2], []));
},
call$4: function($0, $1, $2, $3) {
  return this.noSuchMethod$1(this, H.createInvocationMirror("call", "call$4", 0, [$0, $1, $2, $3], []));
},
call$2$runGuarded: function($0, $1) {
  return this.noSuchMethod$1(this, H.createInvocationMirror("call", "call$2$runGuarded", 0, [$0, $1], ["runGuarded"]));
},
```

In reality, how often are people catching a `NoSuchMethodError`? My guess is close to _0_.

## Proposals

_NOTE: This can wait until the Kernel frontend is complete_

### Option 1: Completely fail if `noSuchMethod` is found anywhere

This might be the most difficult to do, but would eliminate needing any other solutions.

For example, this would be illegal and should fail:

```dart
abstract class A {
  a();
}

class B implements A {
  noSuchMethod(_) => print('Never implemented anything');
}
```

Because we know `noSuchMethod` is never actually implemented then backends could also just segfault/fail if an attempt is made to call an undefined method. For example, I'm imagining this in the `dart2js` backend:

```js
// Psuedo-dart2js referencing the Dart code above
function main() {
  var b = new B();
  b.a(); // <-- Static error if an analyzer flag is enabled, otherwise runtime "undefined"
}
```

### Option 2: Support `noSuchMethod` but in a bare-minimum way

This might be useful to have a phased approach to Option 1.

```js
call$1: function() {
  throw new $dartNoSuchMethodError();
},
call$2: function() {
  throw new $dartNoSuchMethodError();
},
call$0: function() {
  throw new $dartNoSuchMethodError();
},
call$3: function() {
  throw new $dartNoSuchMethodError();
},
```

We still throw a `NoSuchMethodError` (which can be `try-caught`), but there is no longer invocation mirrors in production code - which will probably help a lot. The other advantage is we can strip all mirrors-code that is used to support `noSuchMethod`.

---

Open to other suggestions. This is just a flow of consciousness dump

**Comment by irfantusneem** [comment]
I prefer Option 2.

**Comment by hterkelsen** [comment]
@rakudrama who I think implemented the `noSuchMethod` stubs on `Object`

**Comment by sgonyea** [comment]
I like #1 because it's less work. `noSuchMethod` doesn't add much value when you're compiling to JS.  dart2js should just refuse to compile it.

I'm of the same opinion for operator overloading: I doubt it's something that translates nicely to JS.

**Comment by matanlurey** [comment]
@sgonyea Restrict operator overloading (or any language feature not supported in JS) would definitely cripple the language a bit - I can argue for removing `noSuchMethod` because even in the Dart world you'd probably not want to use it in production code.

On the other hand, being able to overload operators to me makes the language nicer, and the translation to JS doesn't have to be any more complicated than something like:

```dart
class A{
  final int _value;
  
  const A(this._value);

  operator+ (A a) {
    return new A(_value + a._value);
  }
}

add(a, b) {
  return a + b;
}
```

```js
class A {
  constructor(value) {
    this._value = value;
  }

  $add(a) {
    return new A(this._value + a._value);
  }
}

function add(a, b) {
  return a.$add(b);
}
```

**Comment by rakudrama** [comment]
TL;DR: If you don't want the huge dispatch table, don't use noSuchMethod. There is a 'hint' when you do, to give you the control you need. Go and fix your warnings. You might need `--show-package-warnings` to see them.

noSuchMethod _is_ useful if you don't want to fully implement an interface.

We are OK with breaking noSuchMethod for Object.noSuchMethod (which includes Null.noSuchMethod).
By 'breaking' I mean getting a JavaScript TypeError that from which it is impossible to tell the values of the receiver or arguments. If you catch the error, it is converted to a Dart NoSuchMethodError with missing information. This is what happens today if no class implements noSuchMethod.

If every implementation of noSuchMethod delegates to the superclass, then eventually it would end up at Object.noSuchMethod.  Since we have already 'broken' Object.noSuchMethod when it is the only implementation, we OK with that being 'broken' if the only implementations delegate up the superclass chain. This is also implemented today.

By delegate I mean exactly this form:

```dart
noSuchMethod(i) => super.noSuchMethod(i);
```

When that is the only form, we omit the support. @hterkelsen is checking that this is working as he intended.

I believe the problem is that some class implemented noSuchMethod and it is not in the form above.

**Comment by matanlurey** [comment]
> TL;DR: If you don't want the huge dispatch table, don't use noSuchMethod. There is a 'hint' when you do, to give you the control you need. Go and fix your warnings. You might need `--show-package-warnings` to see them.

I've legitimately never seen this. I am assuming this is `dart2js` specific, not the analyzer? I imagine we could do something like ban `noSuchMethod` in code not meant for testing only internally, but it might be difficult.

/cc @bwilkerson and @keertip 

> 
> noSuchMethod _is_ useful if you don't want to fully implement an interface.
> 
> We are OK with breaking noSuchMethod for Object.noSuchMethod (which includes Null.noSuchMethod).

So if I understand this correctly, I could use `noSuchMethod` to avoid implementing an entire interface, and have it just become close to `undefined` if I write (and only write):

```dart
noSuchMethod(_) => super.noSuchMethod(_);
```

> By 'breaking' I mean getting a JavaScript TypeError that from which it is impossible to tell the values of the receiver or arguments. If you catch the error, it is converted to a Dart NoSuchMethodError with missing information. This is what happens today if no class implements noSuchMethod.
> 
> If every implementation of noSuchMethod delegates to the superclass, then eventually it would end up at Object.noSuchMethod.  Since we have already 'broken' Object.noSuchMethod when it is the only implementation, we OK with that being 'broken' if the only implementations delegate up the superclass chain. This is also implemented today.
> 
> By delegate I mean exactly this form:
> 
> ```dart
> noSuchMethod(i) => super.noSuchMethod(i);
> ```

So already if that is the only use of `noSuchMethod` we get the optimization and no dispatch table? Did I understand that correctly?

> 
> When that is the only form, we omit the support. @hterkelsen is checking that this is working as he intended.
> 
> I believe the problem is that some class implemented noSuchMethod and it is not in the form above.

**Comment by hterkelsen** [comment]
Yes, if the only implementations of `noSuchMethod` in your code are of the recognized form, no dispatch table will be generated. Also, type inference will improve a lot, for instance if you have `x.foo()` in your code, `dart2js` can say that `x` has to be one of the types that has a `foo` method, or the method will throw right there. If you have a `noSuchMethod` anywhere in your code, `dart2js` will say that `x.foo()` can return anything.

**Comment by matanlurey** [comment]
> Yes, if the only implementations of `noSuchMethod` in your code are of the recognized form, no dispatch table will be generated. Also, type inference will improve a lot, for instance if you have `x.foo()` in your code, `dart2js` can say that `x` has to be one of the types that has a `foo` method, or the method will throw right there. If you have a `noSuchMethod` anywhere in your code, `dart2js` will say that `x.foo()` can return anything.

Thanks! We should document this properly in our dart2js/production guides internally and externally. @kwalrath and @kevmoo - do we have a section appropriate for a subset of this information?

Also see internal bug 34746130

**Comment by matanlurey** [comment]
Any reason not to bring back `@proxy` instead of forcing teams to write:

```dart
noSuchMethod(i) => super.noSuchMethod(i)
```

Seems like we are torturing users a bit to get around something easy to implement in the analyzer?*

_*Not to make that assumption, just easy in my head._

**Comment by kwalrath** [comment]
> Thanks! We should document this properly in our dart2js/production guides internally and externally. @kwalrath and @kevmoo - do we have a section appropriate for a subset of this information?

This kind of info seems like it should go into a troubleshooting page, which could be associated with the [dart2js docs](https://webdev.dartlang.org/tools/dart2js) (or, in time, with a web app deployment guide).

We could split up the dart2js page when we add the troubleshooting guide. The "Helping dart2js generate better code" and "Debugging" sections could easily be their own pages (assuming they're still relevant).

**Comment by fsc8000** [comment]
For native AOT, there is no size overhead of having ```noSuchMethod```. In some cases we can generate faster code if there is only Object.noSuchMethod (and no user-defined version), but I would not expect a big difference.

**Comment by matanlurey** [comment]
> For native AOT, there is no size overhead of having ```noSuchMethod```. In some cases we can generate faster code if there is only Object.noSuchMethod (and no user-defined version), but I would not expect a big difference.

Just curious what _big difference_ could mean here, especially if something is in a hot loop or critical UI component that is rendered many times.

**Comment by fsc8000** [comment]
> Just curious what big difference could mean here, especially if something is in a hot loop or critical UI component that is rendered many times.

To be precise, what I meant with difference was more for applications where there is no single hot loop, but time spent across hundreds of methods.
But yes, it _could_ make a big difference for a specific call-site inside a hot loop, but a few things have to come together: The compiler propagates the receiver type downwards from call-sites to unique methods. This means that after such a call, the type of the receiver is known, if there is no user-defined noSuchMethod (the compiler knows that Object.noSuchMethod is ok, because it always throws an exception). For example, a loop with at least two calls on the same object, where the first call is to a unique method name would be affected. @rakudrama dart2js does something similar, doesn't it?

**Comment by rakudrama** [comment]
On Thu, Jan 26, 2017 at 6:47 PM, Florian Schneider <notifications@github.com
> wrote:

> Just curious what big difference could mean here, especially if something
> is in a hot loop or critical UI component that is rendered many times.
>
> To be precise, what I meant with difference was more for applications
> where there is no single hot loop, but time spent across hundreds of
> methods.
> But yes, it *could* make a big difference for a specific call-site inside
> a hot loop, but a few things have to come together: The compiler propagates
> the receiver type downwards from call-sites to unique methods. This means
> that after such a call, the type of the receiver is known, if there is no
> user-defined noSuchMethod (the compiler knows that Object.noSuchMethod is
> ok, because it always throws an exception). For example, a loop with at
> least two calls on the same object, where the first call is to a unique
> method name would be affected. @rakudrama <https://github.com/rakudrama>
> dart2js does something similar, doesn't it?
>
TL;DR: Yes, but I don't expect you to notice a perf difference with strong
mode or --trust-type-annotations.

​Our whole program type analysis and per-procedure analysis does pretty
much the same. We call it *receiver type strengthening. *I talked about it
at the 2015 Dart Summit (https://www.youtube.com/watch?v=xneWkyKLuk0),
where I give an example.

There are two separate concerns:

1. If every noSuchMethod implementation obviously throws, we can do
receiver type strengthening, and sometimes get better code if we learn
something that is actually useful.
2. If every noSuchMethod implementation follows the super.noSuchMethod
pattern we can eliminate the big table in addition to receiver type
strengthening.  The big table is somewhere between 0.3%-3% of the program
size.

Strong typing can also help the optimizer - strong mode or
--trust-type-annotations eliminates the possibility of hitting a class with
noSuchMethod in many cases.

**Comment by floitschG** [comment]
Option 3: make `noSuchMethod` only work for the members the class implements. This would still make it viable for Mock classes, for example.

That said: I agree with @rakudrama: with `--trust-type-annotations` or strong mode, most calls should be on typed targets, and it becomes much easier to see which members could actually be the target of nsm.

**Comment by matanlurey** [comment]
I'm guessing this is ~mostly covered by the new nSM semantics.
